ui_framework/components/button/button.js

```diff
@@ -0,0 +1,147 @@
+import React, {
+  PropTypes,
+} from 'react';
+
+import classNames from 'classnames';
+import keyMirror from 'keymirror';
+
+import { KuiButtonIcon } from './button_icon/button_icon';
+
+const commonPropTypes = {
+  type: PropTypes.string,
+  testSubject: PropTypes.string,
+  isDisabled: PropTypes.bool,
+  onClick: PropTypes.func,
+  data: PropTypes.any,
+  className: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.array,
+    PropTypes.object,
+  ]),
+};
+
+const getIcon = props => (
+  props.isLoading
+    ? <KuiButtonIcon type={KuiButtonIcon.TYPE.LOADING} />
+    : props.icon
+);
+
+const getClassName = (props, icon) => {
+  const typeToClassNameMap = {
+    [KuiButton.TYPE.BASIC]: 'kuiButton--basic',
+    [KuiButton.TYPE.HOLLOW]: 'kuiButton--hollow',
+    [KuiButton.TYPE.DANGER]: 'kuiButton--danger',
+    [KuiButton.TYPE.PRIMARY]: 'kuiButton--primary',
+  };
+
+  return classNames('kuiButton', props.className, {
+    [typeToClassNameMap[KuiButton.TYPE[props.type]]]: props.type,
+    'kuiButton--iconText': icon,
+  });
+};
+
+const getChildren = (props, icon) => {
+  // We need to wrap the text so that the icon's :first-child etc. seudo-selectors get applied
+  // correctly.
+  const wrappedChildren = props.children ? <span>{props.children}</span> : undefined;
+
+  return props.isIconOnRight
+    ? (
+      <span>
+        {wrappedChildren}
+        {icon}
+      </span>
+    ) : (
+      <span>
+        {icon}
+        {wrappedChildren}
```

Comment:
Do you think we should have a style rule for spaces inside the brackets?  I previously created a rule in that style guide that said no spaces, but it was based on an invalid assumption (my bad).  Just want that to be clear, so if we want to stick with the current eslint "object-curly-spacing" rule, we can add it to the react style guide.  If you think a rule is too restrictive, I'm fine leaving it open.